interfaces/builtin/content.go

```diff
@@ -0,0 +1,141 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+)
+
+type ContentInterface struct{}
+
+func (iface *ContentInterface) Name() string {
+	return "content"
+}
+
+func (iface *ContentInterface) SanitizeSlot(slot *interfaces.Slot) error {
+	if iface.Name() != slot.Interface {
+		panic(fmt.Sprintf("slot is not of interface %q", iface))
+	}
+
+	// check that we have either a read or write path
+	rpath, rok := slot.Attrs["read"].([]interface{})
+	wpath, wok := slot.Attrs["write"].([]interface{})
+	if !rok && !wok {
+		return fmt.Errorf("content must contain a read or write path")
+	}
+	if len(rpath) == 0 && len(wpath) == 0 {
+		return fmt.Errorf("read or write path must be set")
+	}
+
+	// go over both paths
+	paths := rpath
+	paths = append(paths, wpath...)
+	for _, p := range paths {
+		if strings.Contains(p.(string), "..") {
+			return fmt.Errorf("relative path not allowed")
+		}
+	}
+
+	return nil
+}
+
+func (iface *ContentInterface) SanitizePlug(plug *interfaces.Plug) error {
+	if iface.Name() != plug.Interface {
+		panic(fmt.Sprintf("plug is not of interface %q", iface))
+	}
+	target, ok := plug.Attrs["target"].(string)
+	if !ok || len(target) == 0 {
+		return fmt.Errorf("content plug must contain target path")
+	}
+	if strings.Contains(target, "..") {
+		return fmt.Errorf("relative path not allowed")
+	}
+
+	return nil
+}
+
+func (iface *ContentInterface) ConnectedSlotSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityAppArmor, interfaces.SecuritySecComp, interfaces.SecurityDBus, interfaces.SecurityUDev, interfaces.SecurityMount:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+func (iface *ContentInterface) PermanentSlotSnippet(slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+
+	switch securitySystem {
+	case interfaces.SecurityAppArmor, interfaces.SecuritySecComp, interfaces.SecurityDBus, interfaces.SecurityUDev, interfaces.SecurityMount:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+func (iface *ContentInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	contentSnippet := bytes.NewBuffer(nil)
+	dst := plug.Attrs["target"].(string)
+	dst = filepath.Join(plug.Snap.MountDir(), dst)
+
+	// read
+	if readPaths, ok := slot.Attrs["read"].([]interface{}); ok {
+		for _, r := range readPaths {
+			src := filepath.Join(slot.Snap.MountDir(), r.(string))
+			fmt.Fprintf(contentSnippet, "%s %s none bind,ro 0 0\n", src, dst)
```

Comment:
Can you move those to a helper (say `MountEntry()` or even `BindMountEntry()` so that this is less easy to break.

Offtopic: I was thinking of expanding snippets to be more than `[]byte` as long as it can "render" to []byte so that we can start introducing internal state and APIs for a snippet being constructed. This is better than emitting text and parsing text elsewhere to process it. Before that can happen though I think it is easy to introduce helpers that add help build the snippet as it exist today, this can be done for dbus apparmor rules and a few other things. The end result (policy) is the same, but the code should be easier to read and maintain.